File: test/org/newtonpolyhedron/solve/matrixminorgcd/MatrizMinorGCDSolverImplTest.java
package org.newtonpolyhedron.solve.matrixminorgcd;

import static org.junit.Assert.*;
import static org.newtonpolyhedron.test.TestSupport.*;
import static org.newtonpolyhedron.utils.MatrixUtils.*;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.lang3.tuple.Pair;
import org.junit.Test;

public class MatrizMinorGCDSolverImplTest {
	
	private static final MatrixMinorGCDSolverImpl	solver	= new MatrixMinorGCDSolverImpl();
	
	@Test
	public void case1() throws Exception {
		int[][] source = {//
		//
			{1, 3, 4},//
			{3, 4, 2},//
			{0, 0, 0},//
		};
		int expectedGcd = 5;
		List <BigInteger> expectedMinors = Arrays.asList( //
				BigInteger.valueOf(-10), //
				BigInteger.valueOf(-10), //
				BigInteger.valueOf(-5) //
		);
		Pair <Integer, List <BigInteger>> actual = solver.getLastRowGcd(fromInt(source));
		assertEquals(new Integer(expectedGcd), actual.getLeft());
		assertEquals(expectedMinors, actual.getRight());
	}
	
	@Test
	public void case2() throws Exception {
		int[][] source = {//
		//
			{-3, -2, 1},//
			{5, -2, 1},//
			{0, 0, 0},//
		};
		int expectedGcd = 8;
		List <BigInteger> expectedMinors = Arrays.asList( //
				BigInteger.valueOf(0), //
				BigInteger.valueOf(-8), //
				BigInteger.valueOf(16) //
		);
		Pair <Integer, List <BigInteger>> actual = solver.getLastRowGcd(fromInt(source));
		assertEquals(new Integer(expectedGcd), actual.getLeft());
		assertEquals(expectedMinors, actual.getRight());
	}
	
	@Test
	public void case3() throws Exception {
		int[][] source = {//
		//
			{1, 3, 5, 1},//
			{1, 2, -6, 9},//
			{2, 4, 3, 2},//
			{0, 0, 0, 0},//
		};
		int expectedGcd = 1;
		List <BigInteger> expectedMinors = Arrays.asList( //
				BigInteger.valueOf(73), //
				BigInteger.valueOf(56), //
				BigInteger.valueOf(16), //
				BigInteger.valueOf(-15) //
		);
		Pair <Integer, List <BigInteger>> actual = solver.getLastRowGcd(fromInt(source));
		assertEquals(new Integer(expectedGcd), actual.getLeft());
		assertEquals(expectedMinors, actual.getRight());
	}
}

File: test/org/newtonpolyhedron/solve/matrixminorgcd/MatrizMinorGCDSolverImplTest.scala
package org.newtonpolyhedron.solve.matrixminorgcd
import org.junit.runner.RunWith
import org.newtonpolyhedron.test._
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class MatrizMinorGCDSolverImplTest extends FunSuite {

  val solver = new MatrixMinorGCDSolverImpl

  test("3x3 case 1") {
    val matrix = matrFrac(a(
      a(1, 3, 4),
      a(3, 4, 2),
      a(0, 0, 0)))
    val expectedGcd = 5
    val expectedMinors = Seq(-10, -10, -5) map BigInt.apply

    val res = solver.lastRowGcd(matrix)
    assert(res._1 === expectedGcd)
    assert(res._2 === expectedMinors)
  }

  test("3x3 case 2") {
    val matrix = matrFrac(a(
      a(-3, -2, 1),
      a(5, -2, 1),
      a(0, 0, 0)))
    val expectedGcd = 8
    val expectedMinors = Seq(0, -8, 16) map BigInt.apply

    val res = solver.lastRowGcd(matrix)
    assert(res._1 === expectedGcd)
    assert(res._2 === expectedMinors)
  }

  test("4x4") {
    val matrix = matrFrac(a(
      a(1, 3, 5, 1),
      a(1, 2, -6, 9),
      a(2, 4, 3, 2),
      a(0, 0, 0, 0)))
    val expectedGcd = 1
    val expectedMinors = Seq(73, 56, 16, -15) map BigInt.apply

    val res = solver.lastRowGcd(matrix)
    assert(res._1 === expectedGcd)
    assert(res._2 === expectedMinors)
  }
}